File: questPresentation/src/main/java/ru/nekit/android/qls/quest/answer/CoinQuestAnswerChecker.java
package ru.nekit.android.qls.quest.answer;

import android.support.annotation.NonNull;

import ru.nekit.android.qls.quest.Quest;
import ru.nekit.android.qls.quest.answer.common.QuestAnswerChecker;
import ru.nekit.android.qls.quest.resources.collections.CoinVisualResourceCollection;
import ru.nekit.android.qls.quest.types.NumberSummandQuest;

public class CoinQuestAnswerChecker extends QuestAnswerChecker<CoinVisualResourceCollection> {

    @Override
    public boolean checkAlternativeInput(@NonNull Quest inQuest, @NonNull CoinVisualResourceCollection answer) {
        NumberSummandQuest quest = (NumberSummandQuest) inQuest;
        switch (inQuest.getQuestionType()) {

            case UNKNOWN_MEMBER:

                return CoinVisualResourceCollection.getById(quest.getAnswer()).nomination == answer.nomination;

            default:

                return true;
        }
    }

    @Override
    public boolean checkStringInput(@NonNull Quest inQuest, @NonNull String value) {
        NumberSummandQuest quest = (NumberSummandQuest) inQuest;
        int answer = 0;
        for (int i = 0; i < quest.leftNode.length; i++) {
            answer += CoinVisualResourceCollection.getById(quest.leftNode[i]).nomination;
        }
        return answer == Integer.valueOf(value);
    }
}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/answerChecker/CoinQuestAnswerChecker.kt
package ru.nekit.android.qls.domain.answerChecker

import ru.nekit.android.qls.domain.answerChecker.common.QuestAnswerChecker
import ru.nekit.android.qls.domain.model.quest.NumberSummandQuest
import ru.nekit.android.qls.domain.model.resources.CoinVisualResourceCollection
import ru.nekit.android.qls.shared.model.QuestionType

class CoinQuestAnswerChecker : QuestAnswerChecker<NumberSummandQuest>() {

    override fun checkAlternativeInput(quest: NumberSummandQuest,
                                       answer: Any): Boolean {
        return when (quest.questionType) {

            QuestionType.UNKNOWN_MEMBER ->
                CoinVisualResourceCollection.getById(quest.answer).nomination ==
                        (answer as CoinVisualResourceCollection).nomination

            else ->
                true
        }
    }

    override fun checkStringInput(quest: NumberSummandQuest, value: String): Boolean =
            Integer.valueOf(value) == quest.leftNode.sumBy {
                CoinVisualResourceCollection.getById(it).nomination
            }

}